From 3f66f3aaf74f5e3508544764148d0fd8fcf96976 Mon Sep 17 00:00:00 2001
From: Gerd Hoffmann <kraxel@redhat.com>
Date: Thu, 21 Nov 2024 16:35:41 +0100
Subject: [PATCH 2/2] update pci_pad_mem64 handling

RH-Author: Gerd Hoffmann <kraxel@redhat.com>
RH-MergeRequest: 15: backport opt/org.seabios/pci64 config option
RH-Jira: RHEL-240025
RH-Acked-by: Luigi Leonardi <None>
RH-Acked-by: Oliver Steffen <osteffen@redhat.com>
RH-Commit: [2/2] 38117b77e5e5b83f7cbe1175f916f6205123a920

Add a new possible state: '-1' means 'use default'.  In that case
seabios continue to use the current heuristic:  In case memory
above 4G is present enable 64-bit guest friendly configuration.

This allows forcing the one or the other behavior by setting the
pci_pad_mem64 variable beforehand (which is done by another patch).

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>
(cherry picked from commit df9dd418b3b0e586cb208125094620fc7f90f23d)
---
 src/fw/pciinit.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/src/fw/pciinit.c b/src/fw/pciinit.c
index bb44dc29..658d806d 100644
--- a/src/fw/pciinit.c
+++ b/src/fw/pciinit.c
@@ -55,7 +55,7 @@ u64 pcimem64_end   = BUILD_PCIMEM64_END;
 // Resource allocation limits
 static u64 pci_io_low_end = 0xa000;
 static u64 pci_mem64_top  = 0;
-static u32 pci_pad_mem64  = 0;
+static u32 pci_pad_mem64  = -1;
 
 struct pci_region_entry {
     struct pci_device *dev;
@@ -1197,8 +1197,15 @@ pci_setup(void)
         }
     }
 
-    if (CPUPhysBits >= 36 && CPULongMode && RamSizeOver4G)
-        pci_pad_mem64 = 1;
+    pci_pad_mem64 = romfile_loadbool("opt/org.seabios/pci64", -1);
+
+    if (pci_pad_mem64 == -1)
+        // when not set enable in case memory over 4G is present
+        pci_pad_mem64 = RamSizeOver4G ? 1 : 0;
+
+    if (!CPULongMode)
+        // force off for 32-bit CPUs
+        pci_pad_mem64 = 0;
 
     dprintf(1, "=== PCI bus & bridge init ===\n");
     if (pci_probe_host() != 0) {
-- 
2.52.0

